CLAUDE.md

```diff
@@ -0,0 +1,760 @@
+        │  • forkchoiceUpdate                    │
+        │  • newPayload / getPayload             │
+        └──────────────┬─────────────────────────┘
+                       │ communicates with
+                       ▼
+                External EL Client
+                (Geth/Reth/etc.)
+
+┌───────────────────────────────────────────────────────────────┐
+│                      Storage Backend                          │
+│                 (Shared data layer - DI injected)             │
+│  ┌────────────┐ ┌───────────┐ ┌─────────────────┐ ┌────────┐  │
+│  │ BlockStore │ │ BeaconDB  │ │AvailabilityStore│ │Deposit │  │
+│  │            │ │ (StateDB) │ │ (Blob storage)  │ │Store   │  │
+│  └────────────┘ └───────────┘ └─────────────────┘ └────────┘  │
+└────────────────────────┬──────────────────────────────────────┘
+                         │ accessed by
+     ┌───────────────────┼───────────────────┐
+     │                   │                   │
+     ▼                   ▼                   ▼
+┌─────────────┐ ┌─────────────────┐ ┌───────────────┐
+│ Node API    │ │ Validator       │ │ Blockchain    │
+│ Server      │ │ Service         │ │ Service       │
+│(REST/HTTP)  │ │                 │ │               │
+└─────────────┘ └─────────────────┘ └───────────────┘
+
+Async Background Process:
+┌────────────────────────────────────────────────────┐
+│            Deposit Monitoring Flow                 │
+│  Execution Layer → Deposit Contract → DepositStore │
+└────────────────────────────────────────────────────┘
+```
+
+**System Flow:**
+1. **Service Registry** manages the lifecycle of all services in dependency order
+2. **CometBFT Service** drives the entire system through ABCI callbacks
+3. **Block Building**: CometBFT → PrepareProposal → Validator Service
+4. **Block Processing**: CometBFT → ProcessProposal/FinalizeBlock → Blockchain Service
+5. **State Updates**: Blockchain Service → State Processor → Storage Backend
+6. **External Queries**: Node API → Storage Backend (independent of consensus flow)
+
+**Key Facts:**
+- Service Registry orchestrates startup/shutdown in correct order
+- CometBFT is the consensus driver, not a middle layer
+- Storage Backend is a passive resource, not an active service
+- Node API runs independently, only queries storage
+- Execution Engine communicates with external EL clients
+- Services must wait for dependencies (e.g., EngineClient blocks until EL is ready)
+
+### Core Components
+
+**beacon/** - Consensus layer implementation
+- `blockchain/`: Core service handling block processing, state transitions, fork choice
+  - Integrates with execution engine via Engine API
+  - Manages block and blob storage
+  - Handles deposit processing from execution layer
+- `validator/`: Block building and validation
+  - Payload building coordination with execution client
+  - Blob sidecar creation and bundling
+  - Block proposal generation for CometBFT
+- `payload-time/`: Timing utilities for payload handling
```

Comment:
just decide next block time depending on latest block time and whether block is built optimistically or just in time